src/cdk/tree/nested-node.ts

```diff
@@ -0,0 +1,91 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {
+  AfterContentInit,
+  ContentChildren,
+  Directive,
+  OnDestroy,
+  QueryList,
+} from '@angular/core';
+import {Subject} from 'rxjs/Subject';
+import {takeUntil} from '@angular/cdk/rxjs';
+import {CdkTree} from './tree';
+import {TreeNodeOutlet} from './outlet';
+import {CdkTreeNode} from './node';
+
+/**
+ * Nested node is a child of `<cdk-tree>`. It works with nested tree.
+ * By adding `cdkNestedTreeNode` to the tree node template, children of the parent node will be
+ * added in the `nodeOutlet` in tree node template.
+ * For example:
+ *   <cdk-tree-node cdkNestedTreeNode [cdkNode]="node">
+ *     tree node data: {{node.name}}
+ *     <ng-template nodeOutlet> </ng-template>
+ *   </cdk-tree-node>
+ * The children of node will be automatically added to `cdkNodeOutlet`, the result dom will be like
+ * this:
+ *   <cdk-tree-node cdkNestedTreeNode [cdkNode]="node">
+ *     tree node data: {{node.name}}
+ *     <ng-template cdkNodeOutlet>
+ *       <cdk-tree-node cdKNestedTreeNode [cdkNode]="child1"></cdk-tree-node>
+ *       <cdk-tree-node cdKNestedTreeNode [cdkNode]="child2"></cdk-tree-node>
+ *     </ng-template>
+ *   </cdk-tree-node>
+ */
+@Directive({
+  selector: '[cdkNestedTreeNode]'
+})
+export class CdkNestedTreeNode<T> implements AfterContentInit, OnDestroy {
```

Comment:
`CdkNestedTreeNode` isn't template-ized the same way that flat tree nodes are now. Does that limit us in the future if we want to change up the DOM structure of the nested tree?

Reply:
Now `CdkNestedTreeNode` extends `CdkTreeNode`. Like `CdkTreeNode`, users can put anything in `<cdk-nested-tree-node>` as template, but the children nodes will be added to `[cdkTreeNodeOutlet]` inside the nested node template. If there's no `cdkTreeNodeOutlet`, it works the same way as `CdkTreeNode`.